### PR TITLE
Merge all the clean tests into one file and clean up

### PR DESCRIPTION
**Pull request by willkg**
* Moves `test_security.py` tests into `test_clean.py`
* Removes duplicate tests and unhelpful tests
* Adds additional helpful test cases
* Reworks some tests to be easier and run to read by parametrizing them
* Adds comments and adjusts function names to be more helpful

This is a broad cleanup of `clean` tests. To review this, I'd:

1. go through my comments and see if they make sense
2. go through the final `test_clean.py` file and make sure you understand all the things being tested; for tests that aren't clear or readable, we want to fix that

### CODE REVIEW

File: tests/test_clean.py [hunk 1]
```diff
@@ -1,96 +1,210 @@
...
+from bleach import clean
 from bleach.sanitizer import Cleaner
 
 
-def test_empty():
-    assert bleach.clean('') == ''
+def test_clean_idempotent():
+    """Make sure that applying the filter twice doesn't change anything."""
+    dirty = '<span>invalid & </span> < extra http://link.com<em>'
+    assert clean(clean(dirty)) == clean(dirty)
```
Comment (willkg) [inline tests/test_clean.py]: This is the same test--I moved it to the top figuring "this is testing a property of `clean`".

File: tests/test_clean.py [hunk 2]
```diff
@@ -1,96 +1,210 @@
...
 
-    open_comment = '<!-- this is an open comment'
-    assert bleach.clean(open_comment) == ''
-    assert (
-        bleach.clean(open_comment, strip_comments=False) ==
-        '{0!s}-->'.format(open_comment)
-    )
+    with pytest.raises(TypeError) as e:
+        clean(no_type)
+    assert "NoneType" in str(e)
```
Comment (willkg) [inline tests/test_clean.py]: This is the same test--I moved it to the top figuring "this is testing a property of `clean`".

File: tests/test_clean.py [hunk 3]
```diff
@@ -1,96 +1,210 @@
+import os
+
 from html5lib.filters.base import Filter
 import pytest
 
-import bleach
+from bleach import clean
```
Comment (willkg) [inline tests/test_clean.py]: I changed all instances of `bleach.clean` to `clean` figuring there aren't other instances of "clean", so might as well go for some easy-win brevity.

File: tests/test_clean.py [hunk 4]
```diff
@@ -1,96 +1,210 @@
...
+        clean(no_type)
+    assert "NoneType" in str(e)
 
 
-def test_with_comments():
-    text = '<!-- comment -->Just text'
-    assert bleach.clean(text) == 'Just text'
-    assert bleach.clean(text, strip_comments=False) == text
+def test_empty():
+    assert clean('') == ''
```
Comment (willkg) [inline tests/test_clean.py]: Same test--just without the `bleach.clean`.

File: tests/test_clean.py [hunk 5]
```diff
@@ -1,96 +1,210 @@
...
-        bleach.clean('another <em>good</em> tag') ==
+    ),
+
+    (
+        'another <em>good</em> tag',
         'another <em>good</em> tag'
     )
+])
+def test_content_has_allowed_html(data, expected):
+    assert clean(data) == expected
```
Comment (willkg) [inline tests/test_clean.py]: Same test, but parametrized.

File: tests/test_clean.py [hunk 6]
```diff
@@ -1,96 +1,210 @@
...
 
 
-def test_bad_html():
+def test_html_is_lowercased():
     assert (
-        bleach.clean('a <em>fixed tag') ==
-        'a <em>fixed tag</em>'
+        clean('<A HREF="http://example.com">foo</A>') ==
+        '<a href="http://example.com">foo</a>'
```
Comment (willkg) [inline tests/test_clean.py]: I moved this test from the bottom of the file, improved the function name, and changed the test to use allowed-by-default things to reduce the complexity.

File: tests/test_clean.py [hunk 7]
```diff
@@ -1,96 +1,210 @@
...
+        'text<!-- comment -->'
+    ),
+    (
+        'text<!--comment-->',
+        False,
+        'text<!--comment-->'
     )
+])
+def test_comments(data, should_strip, expected):
+    assert clean(data, strip_comments=should_strip) == expected
```
Comment (willkg) [inline tests/test_clean.py]: We had a bunch of comment tests--this merges them all into one parametrized test. Also, I added some additional test cases here around location of whitespace.

File: tests/test_clean.py [hunk 8]
```diff
@@ -1,96 +1,210 @@
...
+    ('<img src="javascript:alert(\'XSS\');">', '&lt;img src="javascript:alert(\'XSS\');"&gt;'),
+
+    # Test with parens
+    ('a <script>safe()</script> test', 'a &lt;script&gt;safe()&lt;/script&gt; test'),
+
+    # Test with braces
+    ('a <style>body{}</style> test', 'a &lt;style&gt;body{}&lt;/style&gt; test'),
+])
+def test_disallowed_tags(data, expected):
+    assert clean(data) == expected
```
Comment (willkg) [inline tests/test_clean.py]: This merges a few tests into one that covers disallowed tags getting escaped.

File: tests/test_clean.py [hunk 9]
```diff
@@ -1,96 +1,210 @@
...
+        [
+            '&lt;script src="http://xx.com/xss.js" xss=""&gt;&lt;/script&gt;',
+            '&lt;script xss="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
+    )
     assert (
-        bleach.clean(text, attributes=ATTRS) ==
-        '<a href="http://xx.com" rel="alternate">xx.com</a>'
+        clean('<script/src="http://xx.com/xss.js"></script>') ==
+        '&lt;script src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
```
Comment (willkg) [inline tests/test_clean.py]: I inlined the `text` argument. I find it's easier to read and reason about when the in and out are next to each other.

File: tests/test_clean.py [hunk 10]
```diff
@@ -1,96 +1,210 @@
...
+            '&lt;script &lt;b="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
+    )
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
+    assert (
+        clean('<script src="http://xx.com/xss.js" <b>') in
+        [
+            '&lt;script src="http://xx.com/xss.js" &lt;b=""&gt;&lt;/script&gt;',
+            '&lt;script &lt;b="" src="http://xx.com/xss.js"&gt;&lt;/script&gt;'
+        ]
```
Comment (willkg) [inline tests/test_clean.py]: The diff in github here is terrible. I merged a couple of tests between `test_clean.py` and `test_security.py`, but otherwise this is untouched.

File: tests/test_clean.py [hunk 11]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+
+    # html5lib tokenizer unescapes character entities, so these would become '
+    # and " which makes it possible to break out of html attributes.
+    #
+    # Verify that clean() doesn't unescape entities.
+    ('&#39;&#34;', '&#39;&#34;'),
 ])
 def test_character_entities(text, expected):
-    assert bleach.clean(text) == expected
+    assert clean(text) == expected
```
Comment (willkg) [inline tests/test_clean.py]: Merged a couple of tests into this one.

File: tests/test_clean.py [hunk 12]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+    # handle disallowed tag that's deep in the tree
+    (
+        '<p><a href="http://example.com/"><img src="http://example.com/"></a></p>',
+        {'tags': ['a', 'p'], 'strip': True},
         '<p><a href="http://example.com/"></a></p>'
-    )
+    ),
+])
+def test_stripping_tags(data, kwargs, expected):
+    assert clean(data, **kwargs) == expected
```
Comment (willkg) [inline tests/test_clean.py]: The tests for stripping tags was kind of all over, so I moved them to a single parametrized test with comments that (I hope) clarify what's getting tested and cover the cases well.

File: tests/test_clean.py [hunk 13]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+
+@pytest.mark.parametrize('data, expected', [
+    (
+        '<scri<script>pt>alert(1)</scr</script>ipt>',
+        'pt&gt;alert(1)ipt&gt;'
+    ),
+    (
+        '<scri<scri<script>pt>pt>alert(1)</script>',
+        'pt&gt;pt&gt;alert(1)'
+    ),
```
Comment (willkg) [inline tests/test_clean.py]: As a side note, the output here is safe, but I don't know exactly how it goes from the `in` to the `out`. This is definitely curious and maybe worth looking into at some point.

File: tests/test_clean.py [hunk 14]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
 
-def test_lowercase_html():
-    """We should output lowercase HTML."""
+def test_href_with_wrong_tag():
     assert (
-        bleach.clean('<EM CLASS="FOO">BAR</EM>', attributes=['class']) ==
-        '<em class="FOO">BAR</em>'
+        clean('<em href="fail">no link</em>') ==
+        '<em>no link</em>'
     )
```
Comment (willkg) [inline tests/test_clean.py]: The diff here is terrible. This test wasn't changed.

File: tests/test_clean.py [hunk 15]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
 
 
-def test_lowercase_html():
-    """We should output lowercase HTML."""
+def test_href_with_wrong_tag():
     assert (
-        bleach.clean('<EM CLASS="FOO">BAR</EM>', attributes=['class']) ==
-        '<em class="FOO">BAR</em>'
+        clean('<em href="fail">no link</em>') ==
+        '<em>no link</em>'
```
Comment (willkg) [inline tests/test_clean.py]: The diff here is terrible. This test wasn't changed.

File: tests/test_clean.py [hunk 16]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+def test_poster_attribute():
+    """Poster attributes should not allow javascript."""
+    tags = ['video']
+    attrs = {'video': ['poster']}
+
+    test = '<video poster="javascript:alert(1)"></video>'
+    assert clean(test, tags=tags, attributes=attrs) == '<video></video>'
+
+    ok = '<video poster="/foo.png"></video>'
+    assert clean(ok, tags=tags, attributes=attrs) == ok
```
Comment (willkg) [inline tests/test_clean.py]: This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.

File: tests/test_clean.py [hunk 17]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+    }
+
+    assert (
+        clean('<img onclick="evil" src="http://example.com/" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="http://example.com/">'
+    )
+    assert (
+        clean('<img onclick="evil" src="http://badhost.com/" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img>'
+    )
```
Comment (willkg) [inline tests/test_clean.py]: This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.

File: tests/test_clean.py [hunk 18]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+        clean('<abbr title=mytitle>myabbr</abbr>') ==
+        '<abbr title="mytitle">myabbr</abbr>'
+    )
+
+
+def test_unquoted_event_handler_attr_value():
+    assert (
+        clean('<a href="http://xx.com" onclick=foo()>xx.com</a>') ==
+        '<a href="http://xx.com">xx.com</a>'
+    )
```
Comment (willkg) [inline tests/test_clean.py]: This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.

File: tests/test_clean.py [hunk 19]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+        clean('<img href="invalid" src="test" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="test">'
+    )
+
+
+def test_unquoted_attr_values_are_quoted():
+    assert (
+        clean('<abbr title=mytitle>myabbr</abbr>') ==
+        '<abbr title="mytitle">myabbr</abbr>'
+    )
```
Comment (willkg) [inline tests/test_clean.py]: This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.

File: tests/test_clean.py [hunk 20]
```diff
@@ -154,83 +269,168 @@ def test_bare_entities(text, expected):
...
+        '<a href="test">test</a>'
+    )
+    assert (
+        clean('<img onclick="evil" src="test" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="test">'
+    )
+    assert (
+        clean('<img href="invalid" src="test" />', tags=IMG, attributes=IMG_ATTR) ==
+        '<img src="test">'
+    )
```
Comment (willkg) [inline tests/test_clean.py]: This test existed elsewhere in the file. I moved it, but it is otherwise unchanged.

File: tests/test_clean.py [hunk 21]
```diff
@@ -305,11 +505,44 @@ def test_attributes_list():
...
+
+    # Unspecified protocols are not allowed
+    (
+        '<a href="http://xx.com">invalid href</a>',
+        {'protocols': ['myprotocol']},
+        '<a>invalid href</a>'
+    )
+])
+def test_uri_value_allowed_protocols(data, kwargs, expected):
+    assert clean(data, **kwargs) == expected
```
Comment (willkg) [inline tests/test_clean.py]: These tests were kind of all over the place, so I merged them into a single parametrized test.

File: tests/test_clean.py [hunk 22]
```diff
@@ -305,11 +505,44 @@ def test_attributes_list():
...
+        {},
+        '<a>xss</a>'
+    ),
+
+    # File protocol is not allowed by default
+    (
+        '<a href="file:///tmp/foo">foo</a>',
+        {},
+        '<a>foo</a>'
+    ),
```
Comment (willkg) [inline tests/test_clean.py]: Previously, this tested `feed:file:///tmp/foo`. I changed that to `file` because that seemed more helpful.

File: tests/test_clean.py [hunk 23]
```diff
@@ -372,73 +605,77 @@ def test_svg_allow_local_href_nonlocal(text, expected):
...
 def test_sarcasm():
     """Jokes should crash.<sarcasm/>"""
-    dirty = 'Yeah right <sarcasm/>'
-    clean = 'Yeah right &lt;sarcasm/&gt;'
-    assert bleach.clean(dirty) == clean
-
+    assert (
+        clean('Yeah right <sarcasm/>') ==
+        'Yeah right &lt;sarcasm/&gt;'
+    )
```
Comment (willkg) [inline tests/test_clean.py]: I inlined the text here to make it easier to read. I also fixed the xfail reason comment.

File: tests/test_clean.py [hunk 24]
```diff
@@ -372,73 +605,76 @@ def test_svg_allow_local_href_nonlocal(text, expected):
...
-                        token['data'][attr] = 'moo'
+    # Convert vertical tab
+    ('1\f23', '1?23'),
 
-                yield token
+    # Convert a bunch of characters in a string
+    ('import y\bose\bm\bi\bt\be\b', 'import y?ose?m?i?t?e?'),
+])
+def test_invisible_characters(data, expected):
+    assert clean(data) == expected
```
Comment (willkg) [inline tests/test_clean.py]: This is the same test, but moved here.

File: tests/test_clean.py [hunk 25]
```diff
@@ -372,73 +605,76 @@ def test_svg_allow_local_href_nonlocal(text, expected):
...
+    # NOTE(willkg): This strips input and expected which makes it easier to
+    # maintain the files. If there comes a time when the input needs whitespace
+    # at the beginning or end, then we'll have to figure out something else.
+    test_data = test_data.strip()
+    expected = expected.strip()
 
-    with pytest.raises(TypeError) as e:
-        bleach.clean(no_type)
-    assert "NoneType" in str(e)
+    assert clean(test_data) == expected
```
Comment (willkg) [inline tests/test_clean.py]: This was moved from `test_security.py` but is otherwise unchanged.

File: tests/test_clean.py [hunk 26]
```diff
@@ -452,3 +688,27 @@ def test_basics(self):
...
+        }
+        TAGS = ['img']
+
+        cleaner = Cleaner(tags=TAGS, attributes=ATTRS, filters=[MooFilter])
+
+        dirty = 'this is cute! <img src="http://example.com/puppy.jpg" rel="nofollow">'
+        assert (
+            cleaner.clean(dirty) ==
+            'this is cute! <img rel="moo" src="moo">'
+        )
```
Comment (willkg) [inline tests/test_clean.py]: This was a separate function, but it uses `Cleaner`, so I figured I'd move it to the `TestCleaner` test class so all the `Cleaner` stuff is one place.

File: tests/test_clean.py (hunk 6 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 7 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 5 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 4 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 8 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py [hunk 27]
```diff
@@ -1,96 +1,210 @@
...
+    ('a <style>body{}</style> test', 'a &lt;style&gt;body{}&lt;/style&gt; test'),
+])
+def test_disallowed_tags(data, expected):
+    assert clean(data) == expected
 
-    text = '<a href="http://xx.com" rel="alternate">xx.com</a>'
-    assert bleach.clean(text) == '<a href="http://xx.com">xx.com</a>'
+
+def test_invalid_char_in_tag():
+    # NOTE(willkg): Two possible outcomes because attrs aren't ordered
```
Comment (g-k) [inline tests/test_clean.py]: Is there any downside to ordering attrs? I can't think of any getters that depend on their order though looking below it matters for escaped tags.

File: tests/test_clean.py (hunk 9 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 10 above)
Comment (g-k) [reply tests/test_clean.py]: 👍

File: tests/test_clean.py (hunk 22 above)
Comment (g-k) [reply tests/test_clean.py]: I think it'd still be useful to have the double protocol case in addition to this one to make sure we aren't just stripping the first invalid protocol.

File: tests/test_clean.py [hunk 28]
```diff
@@ -372,73 +605,76 @@ def test_svg_allow_local_href_nonlocal(text, expected):
     ATTRS = {
         'pattern': ['id', 'href'],
     }
-    assert bleach.clean(text, tags=TAGS, attributes=ATTRS) == expected
+    assert clean(text, tags=TAGS, attributes=ATTRS) == expected
+
+
+def test_weird_strings():
```
Comment (g-k) [inline tests/test_clean.py]: 😄 I figure this test name is copied from somewhere.

File: tests/test_clean.py (hunk 27 above)
Comment (willkg) [reply tests/test_clean.py]: I'm pretty sure the ordering here happens in html5lib. I'll write up an issue to double-check this. It's possible I fixed the issue that required us to test two possibilities.

File: tests/test_clean.py (hunk 22 above)
Comment (willkg) [reply tests/test_clean.py]: I like this idea. I'll look into this in a followup PR.

**Comment by willkg** [comment]
Thank you!
